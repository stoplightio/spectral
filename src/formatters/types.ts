import { IRuleResult } from '../types';

export type Formatter = (results: IRuleResult[]) => string;
