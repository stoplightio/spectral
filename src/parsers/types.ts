import { GetLocationForJsonPath, IParserResult } from '@stoplight/types';

export interface IParser<R extends IParserResult = IParserResult<unknown>> {
  parse(input: string): R;
  getLocationForJsonPath: GetLocationForJsonPath<R>;
  trapAccess<T extends object = object>(obj: T): T;
}
