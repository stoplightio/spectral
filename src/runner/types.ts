import type { DocumentInventory } from '../documentInventory';
import type { Ruleset } from '../ruleset/ruleset';
import { ISpectralDiagnostic } from '../types';

export interface IRunnerInternalContext {
  ruleset: Ruleset;
  documentInventory: DocumentInventory;
  results: ISpectralDiagnostic[];
  promises: Array<Promise<void>>;
}
