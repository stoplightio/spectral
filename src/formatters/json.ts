import { IRuleResult } from '../types';

export const json = (results: IRuleResult[]): string => {
  const outputJson = results.map(result => {
    return {
      code: result.code,
      path: result.path,
      message: result.message,
      severity: result.severity,
      range: result.range,
      source: result.source,
    };
  });
  return JSON.stringify(outputJson, null, '\t');
};
