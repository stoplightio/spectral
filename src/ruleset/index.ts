export * from './mergers';
export { readRuleset, IRulesetReadOptions } from './readRuleset';
export { assertValidRuleset, ValidationError } from './validation';
export { compileExportedFunction, setFunctionContext } from './utils/evaluators';
export { getDiagnosticSeverity } from './severity';
