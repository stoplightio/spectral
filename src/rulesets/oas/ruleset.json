{
  "rules": {
    "operation-2xx-response": {
      "summary": "Operation must have at least one `2xx` response.",
      "recommended": true,
      "type": "style",
      "given": "$..paths.*[?( @property === 'get' || @property === 'put' || @property === 'post' || @property === 'delete' || @property === 'options' || @property === 'head' || @property === 'patch' || @property === 'trace' )]",
      "then": {
        "field": "responses",
        "function": "oasOp2xxResponse"
      },
      "tags": [
        "operation"
      ]
    },
    "operation-formData-consume-check": {
      "summary": "Operations with an `in: formData` parameter must include `application/x-www-form-urlencoded` or `multipart/form-data` in their `consumes` property.",
      "recommended": true,
      "type": "validation",
      "given": "$..paths.*[?( @property === 'get' || @property === 'put' || @property === 'post' || @property === 'delete' || @property === 'options' || @property === 'head' || @property === 'patch' || @property === 'trace' )]",
      "then": {
        "function": "oasOpFormDataConsumeCheck"
      },
      "tags": [
        "operation"
      ]
    },
    "operation-operationId-unique": {
      "summary": "Every operation must have a unique `operationId`.",
      "recommended": true,
      "type": "validation",
      "severity": 0,
      "given": "$",
      "then": {
        "function": "oasOpIdUnique"
      },
      "tags": [
        "operation"
      ]
    },
    "operation-parameters": {
      "summary": "Operation parameters are unique and non-repeating.",
      "recommended": true,
      "type": "validation",
      "given": "$",
      "then": {
        "function": "oasOpParams"
      },
      "tags": [
        "operation"
      ]
    },
    "path-params": {
      "summary": "Path parameters are correct and valid.",
      "message": "{{error}}",
      "type": "validation",
      "severity": 0,
      "recommended": true,
      "given": "$",
      "then": {
        "function": "oasPathParam"
      },
      "tags": [
        "given"
      ]
    },
    "contact-properties": {
      "summary": "Contact object should have `name`, `url` and `email`.",
      "recommended": false,
      "type": "style",
      "given": "$.info.contact",
      "then": [
        {
          "field": "name",
          "function": "truthy"
        },
        {
          "field": "url",
          "function": "truthy"
        },
        {
          "field": "email",
          "function": "truthy"
        }
      ],
      "tags": [
        "api"
      ]
    },
    "example-value-or-externalValue": {
      "summary": "Example should have either a `value` or `externalValue` field.",
      "recommended": false,
      "type": "style",
      "given": "$..example",
      "then": {
        "function": "xor",
        "functionOptions": {
          "properties": [
            "externalValue",
            "value"
          ]
        }
      }
    },
    "info-contact": {
      "summary": "Info object should contain `contact` object.",
      "recommended": true,
      "type": "style",
      "given": "$",
      "then": {
        "field": "info.contact",
        "function": "truthy"
      },
      "tags": [
        "api"
      ]
    },
    "info-description": {
      "summary": "OpenAPI object info `description` must be present and non-empty string.",
      "recommended": true,
      "type": "style",
      "given": "$",
      "then": {
        "field": "info.description",
        "function": "truthy"
      },
      "tags": [
        "api"
      ]
    },
    "info-license": {
      "summary": "OpenAPI object info `license` must be present and non-empty string.",
      "recommended": false,
      "type": "style",
      "given": "$",
      "then": {
        "field": "info.license",
        "function": "truthy"
      },
      "tags": [
        "api"
      ]
    },
    "license-url": {
      "summary": "License object should include `url`.",
      "recommended": false,
      "type": "style",
      "given": "$",
      "then": {
        "field": "info.license.url",
        "function": "truthy"
      },
      "tags": [
        "api"
      ]
    },
    "no-eval-in-markdown": {
      "summary": "Markdown descriptions should not contain `eval(`.",
      "recommended": false,
      "type": "style",
      "given": "$..*",
      "then": [
        {
          "field": "description",
          "function": "pattern",
          "functionOptions": {
            "notMatch": "eval\\("
          }
        },
        {
          "field": "title",
          "function": "pattern",
          "functionOptions": {
            "notMatch": "eval\\("
          }
        }
      ]
    },
    "no-script-tags-in-markdown": {
      "summary": "Markdown descriptions should not contain `<script>` tags.",
      "recommended": true,
      "type": "style",
      "given": "$..*",
      "then": [
        {
          "field": "description",
          "function": "pattern",
          "functionOptions": {
            "notMatch": "<script"
          }
        },
        {
          "field": "title",
          "function": "pattern",
          "functionOptions": {
            "notMatch": "<script"
          }
        }
      ]
    },
    "only-local-references": {
      "summary": "References should start with `#/`.",
      "recommended": false,
      "type": "style",
      "given": "$..['$ref']",
      "then": {
        "function": "pattern",
        "functionOptions": {
          "match": "^#\\/"
        }
      },
      "tags": [
        "references"
      ]
    },
    "openapi-tags-alphabetical": {
      "summary": "OpenAPI object should have alphabetical `tags`.",
      "recommended": false,
      "type": "style",
      "given": "$",
      "then": {
        "field": "tags",
        "function": "alphabetical",
        "functionOptions": {
          "keyedBy": "name"
        }
      },
      "tags": [
        "api"
      ]
    },
    "openapi-tags": {
      "summary": "OpenAPI object should have non-empty `tags` array.",
      "recommended": false,
      "type": "style",
      "given": "$",
      "then": {
        "field": "tags",
        "function": "truthy"
      },
      "tags": [
        "api"
      ]
    },
    "operation-default-response": {
      "summary": "Operations must have a default response.",
      "recommended": false,
      "type": "style",
      "given": "$..paths.*.*.responses",
      "then": {
        "field": "default",
        "function": "truthy"
      },
      "tags": [
        "operation"
      ]
    },
    "operation-description": {
      "summary": "Operation `description` must be present and non-empty string.",
      "recommended": true,
      "type": "style",
      "given": "$..paths.*[?( @property === 'get' || @property === 'put' || @property === 'post' || @property === 'delete' || @property === 'options' || @property === 'head' || @property === 'patch' || @property === 'trace' )]",
      "then": {
        "field": "description",
        "function": "truthy"
      },
      "tags": [
        "operation"
      ]
    },
    "operation-operationId": {
      "summary": "Operation should have an `operationId`.",
      "recommended": true,
      "type": "style",
      "given": "$..paths.*[?( @property === 'get' || @property === 'put' || @property === 'post' || @property === 'delete' || @property === 'options' || @property === 'head' || @property === 'patch' || @property === 'trace' )]",
      "then": {
        "field": "operationId",
        "function": "truthy"
      },
      "tags": [
        "operation"
      ]
    },
    "operation-operationId-valid-in-url": {
      "summary": "operationId may only use characters that are valid when used in a URL.",
      "recommended": true,
      "type": "validation",
      "given": "$..paths.*[?( @property === 'get' || @property === 'put' || @property === 'post' || @property === 'delete' || @property === 'options' || @property === 'head' || @property === 'patch' || @property === 'trace' )]",
      "then": {
        "field": "operationId",
        "function": "pattern",
        "functionOptions": {
          "match": "^[A-Za-z0-9-._~:/?#\\[\\]@!\\$&'()*+,;=]*$"
        }
      },
      "tags": [
        "operation"
      ]
    },
    "operation-singular-tag": {
      "summary": "Operation may only have one tag.",
      "recommended": false,
      "type": "style",
      "given": "$..paths.*[?( @property === 'get' || @property === 'put' || @property === 'post' || @property === 'delete' || @property === 'options' || @property === 'head' || @property === 'patch' || @property === 'trace' )]",
      "then": {
        "field": "tags",
        "function": "length",
        "functionOptions": {
          "max": 1
        }
      },
      "tags": [
        "operation"
      ]
    },
    "operation-summary-formatted": {
      "summary": "Operation `summary` should start with upper case and end with a dot.",
      "recommended": false,
      "type": "style",
      "given": "$..paths.*[?( @property === 'get' || @property === 'put' || @property === 'post' || @property === 'delete' || @property === 'options' || @property === 'head' || @property === 'patch' || @property === 'trace' )]",
      "then": {
        "field": "summary",
        "function": "pattern",
        "functionOptions": {
          "match": "^[A-Z].*\\.$"
        }
      },
      "tags": [
        "operation"
      ]
    },
    "operation-tags": {
      "summary": "Operation should have non-empty `tags` array.",
      "recommended": true,
      "type": "style",
      "given": "$..paths.*[?( @property === 'get' || @property === 'put' || @property === 'post' || @property === 'delete' || @property === 'options' || @property === 'head' || @property === 'patch' || @property === 'trace' )]",
      "then": {
        "field": "tags",
        "function": "truthy"
      },
      "tags": [
        "operation"
      ]
    },
    "parameter-description": {
      "summary": "Parameter objects should have a `description`.",
      "recommended": false,
      "type": "style",
      "given": "$..parameters[*]",
      "when": {
        "field": "@key",
        "pattern": "^(?!.*(\\$ref))"
      },
      "then": {
        "field": "description",
        "function": "truthy"
      },
      "tags": [
        "parameters"
      ]
    },
    "path-declarations-must-exist": {
      "summary": "given declarations cannot be empty, ex.`/given/{}` is invalid.",
      "recommended": true,
      "type": "style",
      "given": "$..paths",
      "then": {
        "field": "@key",
        "function": "pattern",
        "functionOptions": {
          "notMatch": "{}"
        }
      },
      "tags": [
        "given"
      ]
    },
    "path-keys-no-trailing-slash": {
      "summary": "given keys should not end with a slash.",
      "recommended": true,
      "type": "style",
      "given": "$..paths",
      "then": {
        "field": "@key",
        "function": "pattern",
        "functionOptions": {
          "notMatch": ".+\\/$"
        }
      },
      "tags": [
        "given"
      ]
    },
    "path-not-include-query": {
      "summary": "given keys should not include a query string.",
      "recommended": true,
      "type": "style",
      "given": "$..paths",
      "then": {
        "field": "@key",
        "function": "pattern",
        "functionOptions": {
          "notMatch": "\\?"
        }
      },
      "tags": [
        "given"
      ]
    },
    "tag-description": {
      "summary": "Tag object should have a `description`.",
      "recommended": false,
      "type": "style",
      "given": "$.tags[*]",
      "then": {
        "field": "description",
        "function": "truthy"
      },
      "tags": [
        "api"
      ]
    }
  }
}
