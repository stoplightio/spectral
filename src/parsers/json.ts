import { getLocationForJsonPath, JsonParserResult, parseWithPointers, trapAccess } from '@stoplight/json';
import { IParser } from './types';

export const parseJson = (input: string) =>
  parseWithPointers(input, { ignoreDuplicateKeys: false, preserveKeyOrder: true });

export const Json: IParser<JsonParserResult<unknown>> = {
  parse: parseJson,
  getLocationForJsonPath,
  trapAccess,
};
