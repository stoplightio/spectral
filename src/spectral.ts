import { safeStringify } from '@stoplight/json';
import { Resolver } from '@stoplight/json-ref-resolver';
import { DiagnosticSeverity, Dictionary } from '@stoplight/types';
import { YamlParserResult } from '@stoplight/yaml';
import { memoize, merge } from 'lodash';

import { STATIC_ASSETS } from './assets';
import { Document, IDocument, IParsedResult, isParsedResult, ParsedDocument } from './document';
import { DocumentInventory } from './documentInventory';
import { functions as defaultFunctions } from './functions';
import * as Parsers from './parsers';
import { readRuleset } from './rulesets';
import { compileExportedFunction } from './rulesets/evaluators';
import { mergeExceptions } from './rulesets/mergers/exceptions';
import { IRulesetReadOptions } from './rulesets/reader';
import { DEFAULT_SEVERITY_LEVEL, getDiagnosticSeverity } from './rulesets/severity';
import { runRules } from './runner';
import {
  FormatLookup,
  FunctionCollection,
  IConstructorOpts,
  IResolver,
  IRuleResult,
  IRunOpts,
  ISpectralFullResult,
  PartialRuleCollection,
  RegisteredFormats,
  RuleCollection,
  RunRuleCollection,
} from './types';
import { IRuleset, RulesetExceptionCollection } from './types/ruleset';
import { ComputeFingerprintFunc, defaultComputeResultFingerprint, empty, prepareResults } from './utils';
import { generateDocumentWideResult } from './utils/generateDocumentWideResult';

memoize.Cache = WeakMap;

export * from './types';

export class Spectral {
  private readonly _resolver: IResolver;

  public functions: FunctionCollection = { ...defaultFunctions };
  public rules: RunRuleCollection = {};
  public exceptions: RulesetExceptionCollection = {};
  public formats: RegisteredFormats;

  private readonly _computeFingerprint: ComputeFingerprintFunc;

  constructor(opts?: IConstructorOpts) {
    this._computeFingerprint = memoize(opts?.computeFingerprint || defaultComputeResultFingerprint);
    this._resolver = opts?.resolver || new Resolver();
    this.formats = {};
  }

  public static registerStaticAssets(assets: Dictionary<string, string>) {
    empty(STATIC_ASSETS);
    Object.assign(STATIC_ASSETS, assets);
  }

  public async runWithResolved(
    target: IParsedResult | IDocument | object | string,
    opts: IRunOpts = {},
  ): Promise<ISpectralFullResult> {
    const document: IDocument =
      target instanceof Document
        ? target
        : isParsedResult(target)
        ? new ParsedDocument(target)
        : new Document<unknown, YamlParserResult<unknown>>(
            typeof target === 'string' ? target : safeStringify(target, undefined, 2),
            Parsers.Yaml,
            opts.resolve?.documentUri,
          );

    if (document.source === void 0 && opts.resolve?.documentUri !== void 0) {
      (document as Omit<Document, 'source'> & { source: string }).source = opts.resolve?.documentUri;
    }

    const inventory = new DocumentInventory(document, this._resolver);
    await inventory.resolve();

    const validationResults: IRuleResult[] = [...inventory.diagnostics, ...document.diagnostics, ...inventory.errors];

    if (document.formats === void 0) {
      const registeredFormats = Object.keys(this.formats);
      const foundFormats = registeredFormats.filter(format => this.formats[format](inventory.resolved));
      if (foundFormats.length === 0 && opts.ignoreUnknownFormat !== true) {
        document.formats = null;
        if (registeredFormats.length > 0) {
          validationResults.push(this._generateUnrecognizedFormatError(document));
        }
      } else {
        document.formats = foundFormats;
      }
    }

    return {
      resolved: inventory.resolved,
      results: prepareResults(
        [...validationResults, ...runRules(inventory, this.rules, this.functions, this.exceptions)],
        this._computeFingerprint,
      ),
    };
  }

  public async run(target: IParsedResult | Document | object | string, opts: IRunOpts = {}): Promise<IRuleResult[]> {
    return (await this.runWithResolved(target, opts)).results;
  }

  public setFunctions(functions: FunctionCollection) {
    empty(this.functions);

    Object.assign(this.functions, { ...defaultFunctions, ...functions });
  }

  public setRules(rules: RuleCollection) {
    empty(this.rules);

    for (const name in rules) {
      if (!rules.hasOwnProperty(name)) continue;
      const rule = rules[name];

      this.rules[name] = {
        name,
        ...rule,
        severity: rule.severity === void 0 ? DEFAULT_SEVERITY_LEVEL : getDiagnosticSeverity(rule.severity),
      };
    }
  }

  public mergeRules(rules: PartialRuleCollection) {
    for (const name in rules) {
      if (!rules.hasOwnProperty(name)) continue;
      const rule = rules[name];
      if (rule) {
        this.rules[name] = merge(this.rules[name], rule);
      }
    }
  }

  private setExceptions(exceptions: RulesetExceptionCollection) {
    const target: RulesetExceptionCollection = {};
    mergeExceptions(target, exceptions);

    empty(this.exceptions);

    Object.assign(this.exceptions, target);
  }

  public async loadRuleset(uris: string[] | string, options?: IRulesetReadOptions) {
    this.setRuleset(await readRuleset(Array.isArray(uris) ? uris : [uris], options));
  }

  public setRuleset(ruleset: IRuleset) {
    this.setRules(ruleset.rules);

    this.setFunctions(
      Object.entries(ruleset.functions).reduce<FunctionCollection>(
        (fns, [key, { code, ref, name, source, schema }]) => {
          if (code === void 0) {
            if (ref !== void 0) {
              ({ code } = ruleset.functions[ref]);
            }
          }

          if (code === void 0) {
            // shall we log or sth?
            return fns;
          }

          fns[key] = compileExportedFunction(code, name, source, schema);
          return fns;
        },
        {
          ...defaultFunctions,
        },
      ),
    );

    this.setExceptions(ruleset.exceptions);
  }

  public registerFormat(format: string, fn: FormatLookup) {
    this.formats[format] = fn;
  }

  private _generateUnrecognizedFormatError(document: IDocument): IRuleResult {
    return generateDocumentWideResult(
      document,
      `The provided document does not match any of the registered formats [${Object.keys(this.formats).join(', ')}]`,
      DiagnosticSeverity.Warning,
      'unrecognized-format',
    );
  }
}
