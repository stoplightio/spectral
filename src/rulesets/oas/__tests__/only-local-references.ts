import { DiagnosticSeverity } from '@stoplight/types';
import { RuleType, Spectral } from '../../../spectral';
import * as ruleset from '../ruleset.json';

describe('only-local-references', () => {
  const s = new Spectral();
  s.addRules({
    'only-local-references': Object.assign(ruleset.rules['only-local-references'], {
      enabled: true,
      type: RuleType[ruleset.rules['only-local-references'].type],
    }),
  });

  test('validate a correct object', async () => {
    const results = await s.run({
      swagger: '2.0',
      paths: {},
      parameters: [{ $ref: '#/reference/path' }],
    });
    expect(results).toEqual([
      expect.objectContaining({
        code: 'invalid-ref',
      }),
    ]);
  });

  test('return errors if not local ref', async () => {
    const results = await s.run({
      swagger: '2.0',
      paths: {},
      parameters: [{ $ref: 'https://stoplight.io#/reference/path' }],
    });
    expect(results).toEqual(
      expect.arrayContaining([
        expect.objectContaining({
          code: 'only-local-references',
          message: 'References should start with `#/`.',
          severity: DiagnosticSeverity.Warning,
        }),
      ]),
    );
  });
});
