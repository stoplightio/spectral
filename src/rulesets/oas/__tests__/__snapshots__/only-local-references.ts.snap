// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`only-local-references return errors if not local ref 1`] = `
Array [
  Object {
    "code": "only-local-references",
    "message": "must match the pattern '^#\\\\/'",
    "path": Array [
      "parameters",
      0,
      "$ref",
    ],
    "severity": 1,
  },
]
`;
